drivers/serial/uart_stm32.c

```diff
@@ -803,8 +805,6 @@ static void uart_stm32_isr(const struct device *dev)
 
 #ifdef CONFIG_UART_INTERRUPT_DRIVEN
 	if (LL_USART_IsActiveFlag_TC(UartInstance)) {
-		LL_USART_ClearFlag_TC(UartInstance);
```

Comment:
I've done some quick tests, seems that by just removing this line the shell works as before. It probably requires some more investigation, the IRQ API is really low-level some we've probably missed something.